core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -74,134 +35,346 @@ class Queue[A] private (
-            for {
-              p <- Promise.make[Nothing, A]
-              // add the promise to takers, then:
-              // - try take again in case a value was added since
-              // - wait for the promise to be completed
-              // - clean up resources in case of interruption
-              a <- (IO.sync0 { context =>
-                    takers.offer(p)
-                    unsafeCompleteTakers(context)
-                  } *> p.await).onInterrupt(removeTaker(p))
-            } yield a
-    } yield a
+  def take: IO[E, B]
 
   /**
    * Removes all the values in the queue and returns the list of the values. If the queue
    * is empty returns empty list.
    */
-  final val takeAll: IO[Nothing, List[A]] =
-    for {
-      _ <- checkShutdownState
-
-      as <- IO.sync0 { context =>
-             val as = unsafePollAll(queue)
-             strategy.unsafeOnQueueEmptySpace(queue, context)
-             as
-           }
-    } yield as
+  def takeAll: IO[E, List[B]]
 
   /**
    * Takes up to max number of values in the queue.
    */
-  final def takeUpTo(max: Int): IO[Nothing, List[A]] =
-    for {
-      _ <- checkShutdownState
+  def takeUpTo(max: Int): IO[E, List[B]]
+
+  def map[C](f: B => C): Queue2[E, A, C] =
+    new Queue2[E, A, C] {
+      def capacity: Int                             = self.capacity
+      def offer(a: A): IO[E, Boolean]               = self.offer(a)
+      def offerAll(as: Iterable[A]): IO[E, Boolean] = self.offerAll(as)
+      def awaitShutdown: IO[E, Unit]                = self.awaitShutdown
+      def size: IO[E, Int]                          = self.size
+      def shutdown: IO[E, Unit]                     = self.shutdown
+      def take: IO[E, C]                            = self.take.map(f)
+      def takeAll: IO[E, List[C]]                   = self.takeAll.map(_.map(f))
+      def takeUpTo(max: Int): IO[E, List[C]]        = self.takeUpTo(max).map(_.map(f))
+    }
+
+  def mapM[E2 >: E, C](f: B => IO[E2, C]): Queue2[E2, A, C] =
+    new Queue2[E2, A, C] {
+      def capacity: Int                              = self.capacity
+      def offer(a: A): IO[E2, Boolean]               = self.offer(a)
+      def offerAll(as: Iterable[A]): IO[E2, Boolean] = self.offerAll(as)
+      def awaitShutdown: IO[E2, Unit]                = self.awaitShutdown
+      def size: IO[E2, Int]                          = self.size
+      def shutdown: IO[E2, Unit]                     = self.shutdown
+      def take: IO[E2, C]                            = self.take.flatMap(f)
+      def takeAll: IO[E2, List[C]]                   = self.takeAll.flatMap(IO.foreach(_)(f))
+      def takeUpTo(max: Int): IO[E2, List[C]]        = self.takeUpTo(max).flatMap(IO.foreach(_)(f))
+    }
+
+  def zipWith[E2 >: E, A2 <: A, C, D](that: Queue2[E2, A2, C])(f: (B, C) => D): Queue2[E2, A2, D] =
+    new Queue2[E2, A2, D] {
+      def capacity: Int = self.capacity
+
+      // TODO: self.offer orElse that.offer? Do we need tryOffer?
```

Comment:
@jdegoes @ghostdogpr

Reply:
Not sure what to do here, I can't really imagine a use case for it 🤔
Let's see what @jdegoes thinks!

Reply:
Both should receive the `a` imho. Otherwise the whole `zipWith` combinator does not make much sense.

Reply:
That makes sense - so `zipWith` becomes `&&&` and `orElse/|||` should do what I wrote in the comment, right?

Reply:
I think so yes.